drivers/audio/pdm/pdm_nrfx.c

```diff
@@ -0,0 +1,200 @@
+		LOG_ERR("Overflow error when handling event.\n");
+		return;
+	}
+
+	/* If a buffer was requested,
+	 * provide a new one, and alternate the active buffer
+	 */
+	if (p_evt->buffer_requested) {
+		nrfx_err_t result;
+
+		result = nrfx_pdm_buffer_set(&m_next_buffer[0][m_active_buffer],
+						PDM_NRFX_BUFFER_SIZE);
+
+		if (result != NRFX_SUCCESS) {
+			LOG_ERR("Failed to set new buffer, error %d.\n",
+					result);
+		}
+		m_active_buffer = (~m_active_buffer) & 0x01;
+	}
+
+	/* If a buffer has been released,
+	 * save it and submit it to the workqueue for the event handler
+	 */
+	if (p_evt->buffer_released != NULL) {
+		uint16_t i;
+
+		for (i = 0; i < PDM_NRFX_BUFFER_SIZE; i++) {
+			m_buff_released.buffer[i] = p_evt->buffer_released[i];
+		}
+		k_work_submit(&m_buff_released.work);
+	}
+}
+
+
+static int pdm_nrfx_start(struct device *dev)
+{
+	nrfx_err_t result;
+
+	result = nrfx_pdm_buffer_set(&m_next_buffer[0][m_active_buffer],
+					PDM_NRFX_BUFFER_SIZE);
+
+	if (result != NRFX_SUCCESS) {
+		LOG_ERR("Failed to set new buffer, error %d.\n", result);
+	}
+	m_active_buffer = (~m_active_buffer) & 0x01;
+	result = nrfx_pdm_start();
+	if (result == NRFX_ERROR_BUSY) {
+		LOG_ERR("Failed to start PDM sampling, device %s is busy.\n",
+			dev->name);
+		return -EBUSY;
+	}
+	return 0;
+}
+
+
+static int pdm_nrfx_stop(struct device *dev)
+{
+	nrfx_err_t result = nrfx_pdm_stop();
+
+	if (result == NRFX_ERROR_BUSY) {
+		LOG_ERR("Failed to stop PDM sampling, device %s is busy.\n",
+			dev->name);
+		return -EBUSY;
+	}
+	return 0;
+}
+
+
+static int pdm_nrfx_init(struct device *dev)
+{
+	const struct pdm_nrfx_config *config = dev->config_info;
+
+	nrfx_err_t result;
+
+	result = nrfx_pdm_init(&config->initial_config,
+				pdm_nrfx_event_handler);
+
+	if (result != NRFX_SUCCESS) {
+		LOG_ERR("Failed to initialize device: %s\n", dev->name);
+		return -EBUSY;
+	}
+
+	k_work_q_start(&m_pdm_work_q, pdm_stack,
+			K_THREAD_STACK_SIZEOF(pdm_stack),
+			PDM_PRIORITY);
+	k_work_init(&m_buff_released.work, release_buffer);
+
+	m_active_buffer = 0x00;
+
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
+		    nrfx_isr, nrfx_pdm_irq_handler, 0);
+	return 0;
+}
+
+static const struct pdm_driver_api m_pdm_nrfx_api = {
+	.start = pdm_nrfx_start,
+	.stop = pdm_nrfx_stop,
+	.set_data_handler = pdm_nrfx_set_data_handler
+};
+
+static const struct pdm_nrfx_config m_pdm_nrfx_config = {
+	.initial_config = NRFX_PDM_DEFAULT_CONFIG(CLK_PIN, DIN_PIN)
+};
+
+/*
+ * There is only one instance on supported SoCs, so inst is guaranteed
+ * to be 0 if any instance is okay. (We use pdm_0 above, so the driver
+ * is relying on the numeric instance value in a way that happens to
+ * be safe.)
+ *
+ * Just in case that assumption becomes invalid in the future, we use
+ * a BUILD_ASSERT().
+ */
+#define PDM_INIT(inst)							\
+	do {								\
+		BUILD_ASSERT((inst) == 0,				\
+			"multiple instances not supported");		\
+		DEVICE_AND_API_INIT(pdm_0, DT_INST_LABEL(0),		\
```

Comment:
Since only one instance is supported, why use `DT_INST_FOREACH_STATUS_OKAY()` at all?

Reply:
That is very true, I simply saw it from the ADC driver, and copied it.
The driver now simply uses the `DEVICE_AND_API_INIT`.